ssl/ssl_sess.c

```diff
@@ -161,12 +161,12 @@ SSL_SESSION *SSL_get1_session(SSL *ssl)
      * somebody doesn't free ssl->session between when we check it's non-null
      * and when we up the reference count.
      */
-    CRYPTO_w_lock(CRYPTO_LOCK_SSL_SESSION);
+    CRYPTO_THREAD_read_lock(ssl->lock);
```

Comment:
So, the session is "locked" by using `SSL_SESSION_up_ref()`, but the `ssl->session` value could change (e.g. if `ssl` is freed by another thread) so `ssl` needs locking as well (but only for reading since it's not modified),